gantry/simulator/interfaces.cpp

```diff
@@ -76,10 +80,34 @@ static motor_handler::MotorInterruptHandler motor_interrupt(
 static motor_interrupt_driver::MotorInterruptDriver A(motor_queue,
                                                       motor_interrupt,
                                                       motor_interface);
-
 void interfaces::initialize() {}
 
-auto interfaces::get_can_bus() -> can::bus::CanBus& { return canbus; }
+static po::variables_map options{};
+
+void interfaces::initialize_sim(int argc, char** argv) {
+    auto cmdlinedesc = po::options_description(
+        std::string("simulator for the OT-3 gantry ") +
+        ((get_axis_type() == gantry_x) ? "X" : "Y") + " axis");
+    auto envdesc = po::options_description("");
+    cmdlinedesc.add_options()("help,h", "Show this help message.");
+    auto can_arg_xform = can::sim::transport::add_options(cmdlinedesc, envdesc);
+
+    po::store(po::parse_command_line(argc, argv, cmdlinedesc), options);
+    if (options.count("help")) {
+        std::cout << cmdlinedesc << std::endl;
+        std::exit(0);
+    }
+    po::store(po::parse_environment(envdesc, can_arg_xform), options);
+    po::notify(options);
+}
+
+std::shared_ptr<can::bus::CanBus> canbus;
+
+auto interfaces::get_can_bus() -> can::bus::CanBus& {
+    canbus.reset(
+        new can::sim::bus::SimCANBus(can::sim::transport::create(options)));
```

Comment:
the fundamental problem is that the interfaces design really isn't suitable for this. we need a different interface for simulators that lets us pass created objects around rather than relying them being global.

if we keep this design, which we do in this pr, we should initialize our canbus in `initialize` and use it here. that means it needs a global home. that global home in this case is a `shared_ptr`.

Given that it's a shared ptr, making it valid means that we need to `new` up an object and hand control over - so now the shared ptr is in charge of deleting it.

we could also use an `optional<SimCANBus>` i think, but we have a bunch of shared ptrs everywhere so what's one more

Reply:
OK now I understand why we've got smart pointers everywhere now. Makes sense!